domain/service/commandUseCaseResolver.ts

```diff
@@ -1,84 +1,44 @@
-import { promises as fs } from "fs";
-import path from "path";
-import { Context } from "../../types";
+import CodewarsLeaderboardCommand from "../../application/command/codewarsLeaderboardCommand";
+import DontAskToAskCommand from "../../application/command/dontAskToAskCommand";
+import OnlyCodeQuestionsCommand from "../../application/command/onlyCodeQuestionsCommand";
+import { Command, Context } from "../../types";
 import UseCaseNotFound from "../exception/useCaseNotFound";
-import SendMessageToChannelUseCase from "../../application/usecases/sendMessageToChannel/sendMessageToChannelUseCase";
-import MessageRepository from "../repository/messageRepository";
 import ChatService from "./chatService";
-import LoggerService from "./loggerService";
-import ChannelResolver from "./channelResolver";
 import KataService from "./kataService/kataService";
-import SendCodewarsLeaderboardToChannelUseCase from "../../application/usecases/sendCodewarsLeaderboardToChannel/sendCodewarsLeaderboardToChannelUseCase";
+import LoggerService from "./loggerService";
 
 export default class CommandUseCaseResolver {
-  private messageRepository: MessageRepository;
-
-  private chatService: ChatService;
+  private commands: Command[] = [];
 
   private loggerService: LoggerService;
 
-  private channelResolver: ChannelResolver;
-
-  private kataService: KataService;
-
-  private commandMessages: Record<string, string> = {};
-
   constructor({
-    messageRepository,
     chatService,
-    loggerService,
-    channelResolver,
     kataService,
+    loggerService,
   }: {
-    messageRepository: MessageRepository;
     chatService: ChatService;
-    loggerService: LoggerService;
-    channelResolver: ChannelResolver;
     kataService: KataService;
+    loggerService: LoggerService;
   }) {
-    this.messageRepository = messageRepository;
-    this.chatService = chatService;
     this.loggerService = loggerService;
-    this.channelResolver = channelResolver;
-    this.kataService = kataService;
-  }
 
-  private async loadCommands(): Promise<void> {
-    const filePath = path.join(__dirname, "commands.json");
-    const data = await fs.readFile(filePath, "utf-8");
-    this.commandMessages = JSON.parse(data);
+    this.commands.push(
+      new CodewarsLeaderboardCommand(chatService, kataService),
```

Comment:
Não acho necessário estar a passar estas deps. aqui. 
Elas são utilizadas apenas nos comandos, portanto cada comando pode importar as deps. que precisa não ? 

Isto pode ser parte da "aquitectura hexagonal" e eu não estou a par, nesse caso ignora 👍

Reply:
No fundo, como não há um DI container (que eu saiba) no projeto, acabei por optar criar instancias de comandos que vão viver no scope do `commandUseCaseResolver.ts`. Para criar as instancias, preciso de passar as deps necessarias.

A alternativa seria ter constructores vazios e passar os deps no `.execute()` como sugeriste.

Para ser seguro, teria que validar que as deps passadas são as necessárias para cada Command, em jeito de:
```typescript
async execute(context: Context, deps: { chatService: ChatService; kataService: KataService }): Promise<void> {
    if (!deps.chatService || !deps.kataService) {
      throw new Error('Required dependencies not provided');
    }

    const { chatService, kataService } = deps;
    //...
}
```

já que temos classes, mais vale usar.
Se fosse tudo clean FP, aí sim acho que faria sentido a tua sugestão.


Lemme know what you think!
Cheers,


P.S.:

Kiwi aprova! :D

Reply:
O kiwi aprovar efectivament é um ponto muito forte honestamente...
em que ficamos @pxpm @diomonogatari ? aceitamos e merge?

Reply:
```typescript
// dontAskToAskCommand.ts

import chatService from "../../domain/service/chatService";

private chatService chatService;

constructor() {
    this.chatService = chatService; // error
  }

  async execute(context: Context): Promise<void> {
    await this.chatService.sendMessageToChannel(this.message, context.channelId);
  }
``` 
Infelizmente, para termos instancias de qualquer command em algum sítio vamos ter que passar os serviços necessários para o command funcionar.

Com o sistema atual, ou vêm do constructor ou são passados como argumento do `execute()`.
Mas já vi que posso melhorar na mesma.

O `CommandUseCaseResolver` poderá receber a coleção de commands que são possíveis.
Ou seja:
- Criar as instancias necessárias de `ICommands` no index.ts
- Passar as mesmas para o `CommandUseCaseResolver

Algo do estilo:
```typescript
// index.ts
const commands: Command[] = [
    new CodewarsLeaderboardCommand(chatService, kataService),
    new DontAskToAskCommand(chatService),
    new OnlyCodeQuestionsCommand(chatService)
  ]
const useCaseResolver = new CommandUseCaseResolver({
  commands,
  loggerService,
});

// commandUseCaseResolve.ts
constructor({
    commands,
    loggerService,
  }: {
    commands: Command[];
    loggerService: LoggerService;
  }) {
    this.loggerService = loggerService;

    this.commands = commands;
  }
```


@pxpm e @Alf0nso é isto que irei fazer, mas ainda preciso de dar fix nos tests antes de fazer commit deste final cleanup :D

Cheers

Reply:
Boas @diomonogatari aquilo era meio pseudo-code, não podes dar assign de uma interface. Era mais para dar uma ideia geral do que eu estava a pensar. 
Indo mais concretamente á solução que eu imaginava, seria alguma coisa deste gênero: 

**1)** mover o que é relacionado com o discord para um service na `application/`, eg:
```typescript
// application/services/discordService.ts

import ChatService from "../../domain/service/chatService";
import DiscordChatService from "../../infrastructure/service/discordChatService";
import { Client, Intents } from "discord.js";

const discordClient = new Client({
  intents: [Intents.FLAGS.GUILDS, Intents.FLAGS.GUILD_MEMBERS, Intents.FLAGS.GUILD_MESSAGES],
});

export const 
    chatService: ChatService = new DiscordChatService(discordClient),
    client: any = discordClient;
```
**2)** Remover as dependências do construtor dos comandos :
```typescript
// application/command/dontAskToAskCommand.ts

import { Command, Context } from "../../types";
import { chatService } from "../services/discordService";

export default class DontAskToAskCommand implements Command {
  readonly name = "!ja";

  private readonly message: string =
    "Olá! Experimenta fazer a pergunta diretamente e contar o que já tentaste! Sabe mais aqui :point_right: https://dontasktoask.com/pt-pt/";
  
  async execute(context: Context): Promise<void> {
    await chatService.sendMessageToChannel(this.message, context.channelId);
  }
}
```

```typescript
// no index.js

import { chatService, client } from "./application/services/discordService";

// remover os inits que foram movidos para o discordService.ts
```

E depois lá ter os comandos nalgum array:
```javascript
[
    new dontAskToAskCommand(),
    new otherCommand(),
]

quando chamas o `command.execute(context)` ele lá usa as dependências que precisa, importando as instâncias previamente criadas. 

Mais uma vez, leva isto com um grain of salt, não sei se isto vai contra alguma regra da arquitetura 👍